cpp/include/cudf/scalar/scalar.hpp

```diff
@@ -0,0 +1,251 @@
+ */
+class scalar {
+ public:
+  scalar() = default;
+  ~scalar() = default;
+  scalar(scalar&& other) = default;
+  scalar(scalar const& other) = default;
+  scalar& operator=(scalar const& other) = delete;
+  scalar& operator=(scalar&& other) = delete;
+  
+  /**
+   * @brief Construct a new scalar object
+   * 
+   * @note Do not use this constructor directly. Instead, use a factory method
+   * like make_numeric_scalar or make_string_scalar
+   * 
+   * @param type Data type of the scalar
+   * @param is_valid Whether the value held by the scalar is valid
+   * @param stream The CUDA stream to do the allocation in
+   * @param mr The memory resource to use for allocation
+   */
+  scalar(data_type type, bool is_valid = false, cudaStream_t stream = 0,
+      rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource())
+   : _type(type), _is_valid(is_valid, stream, mr)
+  {}
+
+  /**
+   * @brief Returns the scalar's logical value type
+   */
+  data_type type() const noexcept { return _type; }
+
+  /**
+   * @brief Updates the validity of the value
+   * 
+   * @param is_valid true: set the value to valid. false: set it to null
+   */
+  void set_valid(bool is_valid) { _is_valid.set_value(is_valid); }
+
+  /**
+   * @brief Indicates whether the scalar contains a valid value
+   *
+   * @note Using the value when `is_valid() == false` is undefined behaviour
+   * 
+   * @return true Value is valid
+   * @return false Value is invalid/null
+   */
+  bool is_valid() const { return _is_valid.value(); }
+
+  /**
+   * @brief Returns a raw pointer to the validity bool in device memory
+   */
+  bool* validity_data() { return _is_valid.data(); }
+
+ protected:
+  data_type _type{EMPTY};      ///< Logical type of value in the scalar
+  rmm::device_scalar<bool> _is_valid{};  ///< Device bool signifying validity
+};
+
+/**
+ * @brief An owning class to represent a numerical value in device memory
+ * 
+ * @tparam T the data type of the numerical value
+ */
+template <typename T>
+class numeric_scalar : public scalar {
+   static_assert(is_numeric<T>(), "Unexpected non-numeric type.");
+ public:
+  using value_type = T;
+
+  numeric_scalar() = default;
+  ~numeric_scalar() = default;
+  numeric_scalar(numeric_scalar&& other) = default;
+  numeric_scalar(numeric_scalar const& other) = default;
+
+  /**
+   * @brief Construct a new numeric scalar object
+   * 
+   * @param value The initial value of the scalar
+   * @param is_valid Whether the value held by the scalar is valid
+   * @param stream The CUDA stream to do the allocation in
+   * @param mr The memory resource to use for allocation
+   */
+  numeric_scalar(T value, bool is_valid = true, cudaStream_t stream = 0,
+      rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource())
+   : scalar(data_type(experimental::type_to_id<T>()), is_valid, stream, mr)
+   , _data(value, stream, mr)
+  {}
+
+  /**
+   * @brief Set the value of the scalar
+   */
+  void set_value(T value) { 
```

Comment:
Because this is a low-level api, I am not sure I am comfortable with the assumption that setting the value should implicitly set the validity. The use case I see is operations where the result may have useful data in both validity and value. For instance, the result of the expression `null and true` could be `null` or `true` depending on the trinary logic model the consumer is using.

Reply:
The definition of "null" is missing or invalid data. If it's had a value assigned to it, then it's no longer missing or invalid. 

I'm not really following your example.

Reply:
***My concern is out of scope for this PR. Feel free to ignore it.***

Essentially: A value may contain useful information regardless of it's validity.

A simplistic example is logical operations. Consumers may want to consider logical binary operation results invalid if either or both of the arguments are null. Other consumers may want to know the logical result. It's not up to us to decide how `null or true` is interpreted, but we can give both answers with a single scalar. `value=true` `valid=false`. The consumer may then choose to disregard the validity mask, which would result in perfectly valid logical results.

The counter argument would be "just fill nulls with the value you want, dependent on the operation", which may be fine for singular binary operations, but does not extend to compound operations. The result of `a or b or c` cannot be heuristically determined based solely on the validity of the output - the logical result must be computed along side the validity.

I understand these are not use cases we currently support in libcudf, but I like to be reminded that we are making assumptions here.

Reply:
One can always invalidate a cudf::scalar after setting its value...

Reply:
> A value may contain useful information regardless of it's validity.

I think this is the fundamental point that I disagree with.

In Arrow and libcudf, if a value is "null", whatever happens to be in memory is garbage and it's undefined behavior to rely on it to be anything other than garbage.